frontend/dockerfile/instructions/parse_test.go

```diff
@@ -239,3 +255,81 @@ func TestRunCmdFlagsUsed(t *testing.T) {
 	require.IsType(t, c, &RunCommand{})
 	require.Equal(t, []string{"mount"}, c.(*RunCommand).FlagsUsed)
 }
+
+func TestCopyHeredoc(t *testing.T) {
+	cases := []struct {
+		dockerfile    string
+		content       []string
+		preventExpand bool
+	}{
+		{
+			dockerfile: "COPY /foo /bar",
+			content:    nil,
+		},
+		{
+			dockerfile: "COPY <<EOF /bar\nEOF",
+			content:    []string{},
+		},
+		{
+			dockerfile: "COPY <<EOF /bar\nTESTING\nEOF",
+			content:    []string{"TESTING"},
+		},
+		{
+			dockerfile:    "COPY <<'EOF' /bar\nTESTING\nEOF",
+			content:       []string{"TESTING"},
+			preventExpand: true,
+		},
+	}
+
+	for _, c := range cases {
+		r := strings.NewReader(c.dockerfile)
+		ast, err := parser.Parse(r)
+		require.NoError(t, err)
+
+		n := ast.AST.Children[0]
+		comm, err := ParseInstruction(n)
+		require.NoError(t, err)
+		require.Equal(t, c.content, comm.(*CopyCommand).Content)
+		require.Equal(t, c.preventExpand, comm.(*CopyCommand).PreventExpand)
+	}
+}
+
+func TestRunHeredoc(t *testing.T) {
+	cases := []struct {
+		dockerfile string
+		commands   []strslice.StrSlice
+		shell      bool
+	}{
+		{
+			dockerfile: "RUN ls /",
+			commands:   []strslice.StrSlice{{"ls /"}},
+			shell:      true,
+		},
+		{
+			dockerfile: `RUN ["ls", "/"]`,
+			commands:   []strslice.StrSlice{{"ls", "/"}},
+			shell:      false,
+		},
+		{
+			dockerfile: "RUN [\"<<EOF\"]\nls /\nEOF",
```

Comment:
Good catch, that's quite nasty to read at the moment, will fix that up.

Yeah, how heredocs would interact with JSON was never defined anywhere, but likewise, I'm also leaning towards "keep JSON JSON" and don't parse heredocs after this. Amusingly though, I think it's actually *easier* to handle it weirdly like this, but yeah, I'll work out how to make sure JSON doesn't get messed with by heredocs.

Reply:
Thanks! Yes, if possible, producing an error that it's not supported is the "safe" option; we can always decide to _add_ support in future (if we think that's a good thing), but the reverse is a lot more complicated.

The JSON / non-JSON is already confusing many users, e.g. trying to use single quotes `['foo']` (which is not JSON, so Docker will treat it as a "shell" command), so adding heredoc to that makes it even more confusing. Of course (as said above), it would still allow us to change our minds in future.

Reply:
I've reworked the behavior here, so if the line is parsed as JSON, no heredoc is read.

So for example:

```dockerfile
FROM alpine:3.6
RUN ["whoami"]
RUN ["<<EOF"]
RUN ["<<'EOF'"]
```

All of these will be parsed more intuitively now.